gpflow/experimental/check_shapes/checker_context.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2022 The GPflow Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Utilities for accessing the ShapeChecker from a wrapping `check_shapes` decorator.
+"""
+from contextlib import contextmanager
+from contextvars import ContextVar
+from typing import Iterator, Optional
+
+from .checker import S, ShapeChecker, TensorSpecLike
+from .config import get_enable_check_shapes
+from .error_contexts import ErrorContext, FunctionCallContext
+
+_shape_checker: ContextVar[ShapeChecker] = ContextVar("_shape_checker")
+
+
+@contextmanager
+def set_shape_checker(checker: ShapeChecker) -> Iterator[None]:
+    """
+    Sets the current shape checker in the context.
+    """
+    token = _shape_checker.set(checker)
+    try:
+        yield
+    finally:
+        _shape_checker.reset(token)
+
+
+def get_shape_checker() -> ShapeChecker:
+    """
+    Get the :class:`ShapeChecker` from the wrapping :func:`check_shapes` decorator.
+
+    Behaviour is undefined if you call this from a function that is not directly wrapped in
```

Comment:
Why "undefined" rather than "returns a `LookupError`"?

Reply:
> Why "undefined" rather than "returns a `LookupError`"?

Well, what happens depends...
I can do something like:

```python
def g():
    get_shape_checker() 

@check_shapes()
def f():
     g()
```

With my current implementation `g` will see the `ShapeChecker` from `f` which is *weird*. I'd like to avoid that but I don't know a fast/easy way to do that. So, I declare the behaviour undefined, so I'm allowed to fix it in the future if I find a way to do it.